interfaces/hotplug/udevadm.go

```diff
@@ -0,0 +1,150 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package hotplug
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os/exec"
+	"strings"
+)
+
+var udevadmBin = `udevadm`
+
+func parseEnvBlock(block string) (map[string]string, error) {
+	env := make(map[string]string)
+	for i, line := range strings.Split(block, "\n") {
+		if i == 0 && !strings.HasPrefix(line, "P: ") {
+			return nil, fmt.Errorf("no device block marker found before %q", line)
+		}
+		// We are only interested in 'E' properties as they carry all the interesting data,
+		// including DEVPATH and DEVNAME which seem to mirror the 'P' and 'N' values.
+		if strings.HasPrefix(line, "E: ") {
+			if kv := strings.SplitN(line[3:], "=", 2); len(kv) == 2 {
+				env[kv[0]] = kv[1]
+			} else {
+				return nil, fmt.Errorf("failed to parse udevadm output %q", line)
+			}
+		}
+	}
+	return env, nil
+}
+
+func scanDoubleNewline(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	if atEOF && len(data) == 0 {
+		return 0, nil, nil
+	}
+
+	if i := bytes.Index(data, []byte("\n\n")); i >= 0 {
+		// we found data
+		return i + 2, data[0:i], nil
+	}
+
+	// If we're at EOF, return what is left.
+	if atEOF {
+		return len(data), data, nil
+	}
+	// Request more data.
+	return 0, nil, nil
+}
+
+func parseUdevadmOutput(cmd *exec.Cmd, rd *bufio.Scanner, devices chan<- *HotplugDeviceInfo, parseErrors chan<- error) {
+	defer close(devices)
+	defer close(parseErrors)
+
+	env := make(map[string]string)
+
+	for rd.Scan() {
+		// udevadm export output is divided in per-device blocks, blocks are separated
```

Comment:
(nitpick^2) We *could* move this comment now at the beginning of the file as the parsing of udevadm is now split across the file a bit more (before this was the perfect place because all parsing happend below the comment).

Reply:
Done.